Support sample weights in all circa*() functions

ATpoint opened the pull request:
Dear Rex, 

following up late on https://github.com/RWParsons/circacompare/issues/7 I added sample weights support to both `circacompare()`, `circa_single()` and `model_each_group()` (the `nls()` part of it) and added a toy example to the `@examples`. 

I intentionally did not add it to the mixed effect functions as it is unclear to me how. I think it is via something like `nlme(..., weights = varPower(form=~sample_weights))` but nlme documentation and google search did not make me confident enough to add it. 

Right now `circacompare()` and `circa_single()` have a new argument `sample_weights` that write the sample weights to the `nls()` input data, or when weights are NULL (the default) it simply adds a `1` per sample, so no weighting takes place. 

Please have a look if you have time.

best,
Alex

ATpoint committed:
first attempt to support per-sample weights

ATpoint committed:
support sample weights

ATpoint committed:
weights only for nls, not for nlme

ATpoint committed:
fix malformatted authors

RWParsons commented on NEWS.md:
```diff
@@ -1,3 +1,9 @@
+# circacompare 0.1.1.9001
```
Could you please add this under the 0.1.1.9000 heading instead of this new version? There hasn't been an associated change to the version number in the DESCRIPTION file to reflect this dev version.

ATpoint replied:
All changes now under 0.1.1.9000

RWParsons commented on NEWS.md:
```diff
@@ -1,3 +1,9 @@
+
+## Improvements
+
+* allow per-sample weights in `circacompare()` and `circa_single()`. This allows to downweight individual samples, for example in the case of outliers rather than hard-filtered them which reduces power. Such per-sample weights can reproducibly be estimated using approaches such as `arrayWeights()` in the [limma](https://bioconductor.org/packages/release/bioc/html/limma.html) package for datasets  ...
```
I'd prefer not to suggest how people handle outliers. In the documentation or in the version history. There are many reasons why people may want to use weights other than handling outliers and I want to be agnostic to these methods. Could you please reword to state the change performed and not the motivation?

ATpoint replied:
Removed motivation, only stating the change.

RWParsons commented on R/circa_single.R:
```diff
@@ -11,6 +11,7 @@
 #' @param timeout_n The upper limit for the model fitting attempts. Default is 10,000.
 #' @param return_figure Whether or not to return a ggplot graph of the rhythm and cosine model.
 #' @param control \code{list}. Used to control the parameterization of the model.
+#' @param sample_weights A numeric vector of per-sample weights to allow downweighting of outlier samples.
```
Could we please change the argument name to be `weights` to keep it consistent with `nls()` which is being called. I'd also prefer if the argument description was copied from `?nls()`
> `weights` an optional numeric vector of (fixed) weights. When present, the objective function is weighted least squares.

ATpoint replied:
Renamed `sample_weights` to `weights` everwhere.

RWParsons commented on R/circa_single.R:
```diff
@@ -19,7 +20,15 @@
+#' out
+#' 
+#' # with sample weights (arbitrary weights for demonstration)
+#' set.seed(1)
```
No need to set a seed for examples

ATpoint replied:
removed `set.seed()` from all examples

RWParsons commented on R/circa_single.R:
```diff
@@ -81,6 +91,14 @@ circa_single <- function(x,
   }
+  
+  if(!is.null(sample_weights)){
+    l.weights <- length(sample_weights)
```
Could you please change this to `snake_case` to be consistent with the rest of the package? i.e: `l.weights` to become `len_weights`.

ATpoint replied:
done, and also moved the weights validation from the `circa*.R` scripts to `utils.R` to avoid redundancy. Now each `circa*.R` calls `check_weights()` to validate that the weights have same length as the data, are non-negative, numeric and do not contain NAs.

RWParsons commented on R/circacompare.R:
```diff
@@ -11,6 +11,7 @@
 #' @param alpha_threshold The level of alpha for which the presence of rhythmicity is considered. Default is 0.05.
 #' @param timeout_n The upper limit for the model fitting attempts. Default is 10,000.
 #' @param control \code{list}. Used to control the parameterization of the model.
+#' @param sample_weights A numeric vector of per-sample weights to allow downweighting of outlier samples.
```
please make the same change as to `circa_single()` docs and argument name to become `weights` etc.

ATpoint replied:
it's done

RWParsons commented on R/circacompare.R:
```diff
@@ -23,6 +24,16 @@
+#' 
+#' # with sample weights (arbitrary weights for demonstration)
+#' set.seed(1)
+#' sw <- jitter(rep(1, nrow(df)), factor=2)
```
I suggest using getting weights from `runif()`
> sw <- runif(n = nrow(df))

ATpoint replied:
replaced `jitter`+`rep` with `runif`

RWParsons reviewed (changes requested):
Hi Alex - thanks for the PR and new feature. 

I've made some feedback - could you please make those suggested changes?

Also, there are currently no new tests. Could you please add a test or two to ensure that the weights arguments are being used appropriately? For example, you could fit a model and then check that the nls model within contains a weights vector within the object.

```
df <- make_data(phi1 = 6)
out <- circacompare(
    x = df, col_time = "time", col_group = "group",
    col_outcome = "measure"
)
expect_true(all(out$fit$weights) == 1)

sw <- jitter(rep(1, nrow(df)), factor=2)

out2 <- circacompare(
    x = df, col_time = "time", col_group = "group",
    col_outcome = "measure", sample_weights = sw
)

expect_false(all(out2$fit$weights == 1))
```
or something to that effect?

Thank you!!
Rex

RWParsons commented:
Also, I'm pretty sure you're right about the use of `varPower` in `nlme`. I'd be very happy if you could implement that for `circa_single_mixed()` and `circacompare_mixed()`! 

The function documentation could explicitly state that these weights are passed using `varPower `as weights when fitting the model and refer the user to see `?nlme::nlme` for further detail.

Thanks!

ATpoint committed:
move weights check to utils.R

ATpoint committed:
move weights check to utils.R

ATpoint committed:
weights for nlme::nlme

ATpoint committed:
weights must not be negative

ATpoint committed:
support weights in the mixed functions

ATpoint committed:
add tests for weights

ATpoint committed:
modify param description of weights

ATpoint committed:
version bump to 0.1.1.9001

ATpoint committed:
all changes under 0.1.1.9000

ATpoint commented:
Hi Rex,

thanks for the review, good points made. I resolved all requested changes, implemented weights for the mixed functions and added tests for all four `circa*()` functions. Argument description for weights in the mixed functions now mentions that `varPower()` is used to pass the weights to `nlme()` and the tests check that this actually ends up in the fit object (`fit$apVar`). It's all under 0.1.1.9000 in DESCRIPTION/NEWS. 

Please let me know if I missed anything or other changes are necessary.

best,
Alex

ATpoint committed:
fix test

RWParsons committed:
avoid NOTES for long circa_single() doc and weights

RWParsons committed:
avoid inline ifelse statements and run styler

RWParsons commented:
> Hi Rex,
> 
> thanks for the review, good points made. I resolved all requested changes, implemented weights for the mixed functions and added tests for all four `circa*()` functions. Argument description for weights in the mixed functions now mentions that `varPower()` is used to pass the weights to `nlme()` and the tests check that this actually ends up in the fit object (`fit$apVar`). It's all under 0.1.1.9000 in DESCRIPTION/NEWS.
> 
> Please let me know if I missed anything or other changes are necessary.
> 
> best, Alex

Thanks Alex - all your changes look good to me. I've just made a couple commits but these are just formatting changes.
